### PR TITLE
MP-4509 Table Design

### PR DESCRIPTION
**Pull request by danilmoro3**
Ticket: https://spryker.atlassian.net/browse/MP-4509

### CODE REVIEW

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+---
+
+This article describes the Table Design in the Components Library.
+
+## Overview
+
+A Table Component is an arrangement of data in rows and columns, or possibly in a more complex structure 
+(with sorting, filtering, pagination, row selections, infinite scrolling, etc.). 
+It is an essential building block of a user interface.
+A basic able Component is `<spy-table [config]="config"></spy-table>` where `config` is:
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:14]:
```suggestion
A basic table component is `<spy-table [config]="config"></spy-table>` where `config` is:
```

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+- `dataSource` - an array of rows to be displayed on the table.
+- `columns` - an array of columns.
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:17]:
```suggestion
- `dataSource` - Datasource configuration from which the data is taken.
- `columns` - an array of columns configuration.
```

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    { id: 'col3', title: 'Column #3' },
+  ],
+};
+```
+
+## Architecture
+
+![Table Architecture](https://spryker.s3.eu-central-1.amazonaws.com/docs/Marketplace/dev+guides/Front-end/table-architecture.svg)
+
+### Components
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:38]: This section is not clear. What are you trying to explain here?

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+### Features
+
+Every other piece of functionality was extracted into so called `Table Feature`:
+
+- Table Feature is simply an Angular Component that encapsulates specific extension of the Core Table.
+- Core Table contains specific placeholders in it’s view that Table Feature may target to render it’s own piece of UI.
+- Most of the common table functionality already exists as a Table Feature and may be used in project.
+
+To use Features, Table Module must implement a specific interface (TableConfig) and then be registered to the Root Module via TableModule.withFeatures():
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:96]:
```suggestion
To use a Feature component - register an Angular Module that implements `ModuleWithFeature` interface in the Root Module via TableModule.withFeatures() under the key that will become it's configuration key:
```

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        }),
+    ],
+    providers: [
+        {
+            provide: TableDefaultConfig,
+            useClass: TableDefaultConfigData,
+        },
+    ],
+})
+export class DefaultTableConfigModule {}
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:135]: We do not need all features here just one for example

```suggestion
@NgModule({
    imports: [
        TableModule.withFeatures({
            pagination: () => import('@spryker/table.feature.pagination').then((m) => m.TablePaginationFeatureModule),
        }),
    ],
})
export class AppModule {}
```

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+The Table Component is configured via so called [Table Configuration](/docs/marketplace/dev/front-end/table-design/table-configuration.html) 
+that sets up how table will behave and look like.
+
+### Datasources
+
+To render data Table must receive it via so called [Datasources](/docs/marketplace/dev/front-end/ui-components-library/datasources.html) that is registered by the user and then configured via Table Configuration.
+
+### Features
+
+Every other piece of functionality was extracted into so called `Table Feature`:
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:90]: Please also add a link to where all table features are described.

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+})
+export class DefaultTableConfigModule {}
+```
+
+### Columns
+
+Columns in a Table are defined by the `Column Type` and are rendered withing the columns (text, image, link, etc.).
+New Column Type may be created and registered to the Table.
+
+To use Column components, Table Module must implement a specific interface (TableConfig) and then be registered to the Root Module via TableModule.withColumnComponents():
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:143]:
```suggestion
To use Column component it must implement `TableColumn` interface with defined config and then be registered to the Root Module via TableModule.withColumnComponents():
```

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+
+```ts
+class TableDefaultConfigData implements Partial<TableConfig> {
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:146]: Please only keep example of how one column is registered and remove everything else.

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            useClass: TableDefaultConfigData,
+        },
+    ],
+})
+export class DefaultTableConfigModule {}
+```
+
+### Columns
+
+Columns in a Table are defined by the `Column Type` and are rendered withing the columns (text, image, link, etc.).
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:140]: Please add a link to all table columns.

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+}
+```
+
+## Extensions
+
+There are a few common Table extensions that are available in UI library:
+
+- [Column Type](/docs/marketplace/dev/front-end/table-design/table-column-type-extension/) - describes how a specific type of the column is rendered within a table column.
+- [Feature](/docs/marketplace/dev/front-end/table-design/table-feature-extension/) - encapsulating a piece of UI that is targeted to a specific location within a Table Component or that may provide additional functionality.
+- [Filter](/docs/marketplace/dev/front-end/table-design/table-filter-extension/) - provides filtering functionality to the Core Table Component.
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:361]: Please remove this section as everything is described above.

File: docs/marketplace/dev/front-end/table-design/index.md [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+            useClass: TableDefaultConfigData,
+        },
+    ],
+})
+export class DefaultTableConfigModule {}
+```
+
+### Actions 
+
+There is a way to trigger some [Actions](/docs/marketplace/dev/front-end/ui-components-library/actions.html) by the user interaction with the Table.
```
Comment (gund) [inline docs/marketplace/dev/front-end/table-design/index.md:191]: Please add info that there are Table Features that can trigger actions like Row Actions Feature and Batch Actions Feature.

**Review by gund — CHANGES_REQUESTED** [review]

File: docs/marketplace/dev/front-end/table-design/index.md (hunk 3 above)
Comment (danilmoro3) [reply docs/marketplace/dev/front-end/table-design/index.md:38]: This is the main section of table architecture with a diagram and subsections like features, components, etc.

**Review by FRITS-Kh — DISMISSED** [review]

**Review by gibmirmilch — APPROVED** [review]

**Pull request merged**